Use NativePromise chaining in place of operations queue in SourceBufferPrivate

jyavenard opened the pull request:
#### 19836c74aa9f0c3783d5c3f3e9b8b9e3b415afd6
<pre>
Use NativePromise chaining in place of operations queue in SourceBufferPrivate
<a href="https://bugs.webkit.org/show_bug.cgi?id=264854">https://bugs.webkit.org/show_bug.cgi?id=264854</a>
<a href="https://rdar.apple.com/118429088">rdar://118429088</a>

Reviewed by Jer Noble.

By using promise chaining we can remove the need for an operations queue to serialise the tasks.
It also makes the code easier to read/follow as the entire appendBuffer operation is now available at a
glance.
This will simplify making SourceBufferPrivate::append/removeCodedFrames be made asynchronous and use
NativePromise.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::didUpdateFormatDescriptionForTrackId):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::processPendingMediaSamples):
(WebCore::SourceBufferPrivate::processMediaSample):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::~SourceBufferPrivate): Deleted.
(WebCore::SourceBufferPrivate::advanceOperationState): Deleted.
(WebCore::SourceBufferPrivate::rewindOperationState): Deleted.
(WebCore::SourceBufferPrivate::processAppendCompletedOperation): Deleted.
(WebCore::SourceBufferPrivate::queueOperation): Deleted.
(WebCore::SourceBufferPrivate::processPendingOperations): Deleted.
(WebCore::SourceBufferPrivate::abortPendingOperations): Deleted.
(WebCore::SourceBufferPrivate::processError): Deleted.
(WebCore::SourceBufferPrivate::processInitOperation): Deleted.
(WebCore::SourceBufferPrivate::processMediaSamplesOperation): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::didReceiveAllPendingSamples): Test that the promise wasn&apos;t previously resolved.
fallout from bug 264847.
(WebCore::SourceBufferPrivateGStreamer::appendParsingFailed):

Canonical link: <a href="https://commits.webkit.org/270808@main">https://commits.webkit.org/270808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96a3f81fddd28e9c097fb88b6c56803d539f2ee9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27716 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28655 "") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24207 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6883 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2494 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/28655 "") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22732 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/28655 "") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23708 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29170 "Failed to print configuration") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24120 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24107 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/29170 "Failed to print configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3520 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/29170 "Failed to print configuration") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4953 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6354 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3999 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3844 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.h:
```diff
@@ -38,9 +38,12 @@
+#include <wtf/NativePromise.h>
 #include <wtf/RefCounted.h>
 #include <wtf/UniqueRef.h>
+#include <wtf/Vector.h>
```
`optional` and `Vector.h` don't appear to be used in the header. Can these be removed?

jyavenard replied:
they were used, I added them because compilation broke elsewhere following unified build changes.
We have `Vector<PlatformTimeRanges> activeRanges()` method .
and we have `std::optional<SeekTarget> m_pendingSeekTarget;` class member

jernoble commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -245,56 +247,33 @@ void MediaSource::completeSeek()
-        return;
-    }
+    if (isClosed())
+        return GenericPromise::createAndReject(-1);
```
This pattern is repeated a lot; I assume that `-1` will be replaced with a `PlatformMediaError` in a future patch?

jyavenard replied:
Yes... The "PR#4" adds PlatformMediaError and those are replaced everywhere

jernoble commented on Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
```diff
@@ -193,37 +198,38 @@ void SourceBufferPrivateGStreamer::allSamplesInTrackEnqueued(const AtomString& t
-                m_tracks.add(trackInfo.track->id(), MediaSourceTrackGStreamer::create(TrackPrivateBaseGStreamer::TrackType::Text, trackInfo.track->id(), WTFMove(initialCaps)));
-        }
+    SourceBufferPrivate::didReceiveInitializationSegment(WTFMove(initializationSegment));
+}
```
Seems like this could just be removed entirely.

jyavenard replied:
yes indeed.

jernoble commented on Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
```diff
@@ -233,12 +239,14 @@ void SourceBufferPrivateGStreamer::didReceiveSample(Ref<MediaSample>&& sample)
 void SourceBufferPrivateGStreamer::appendParsingFailed()
 {
-    SourceBufferPrivate::appendCompleted(false);
+    m_appendPromise->reject(1);
```
Should this be `-1`? (Or it doesn't matter because this will be a `PlatformMediaError` in the future.)

jernoble commented on Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
```diff
@@ -43,6 +43,7 @@
 #include "TrackPrivateBaseGStreamer.h"
 #include "WebKitMediaSourceGStreamer.h"
 #include <wtf/LoggerHelper.h>
+#include <wtf/NativePromise.h>
```
Looks like the optional.h from above needs to move here.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -129,14 +127,14 @@ void RemoteSourceBufferProxy::sourceBufferPrivateHighestPresentationTimestampCha
+        return GenericPromise::createAndReject(-1);
 
-    m_connectionToWebProcess->connection().sendWithAsyncReply(Messages::SourceBufferPrivateRemote::SourceBufferPrivateDurationChanged(duration), WTFMove(completionHandler), m_identifier);
+    return m_connectionToWebProcess->connection().sendWithPromisedReply(Messages::SourceBufferPrivateRemote::SourceBufferPrivateDurationChanged(duration), m_identifier)->whenSettled(RunLoop::main(), [] {
```
Seems like `sendWithPromisedReply(...)` should have a variant that doesn't take a lambda, and just creates and returns a `GenericPromise` with the result.

jyavenard replied:
`sendWithPromisedReply` doesn't take anything, it returns a NativePromise. You don't have to act on the promise

jyavenard replied:
The issue is returning the IPC error code; this wasn't handled before, and code always assumed that sending an IPC message was successful. We needed a way to pass that error. Merging the error code with the PlatformMediaError would be nice, but I don't know how that will play for consistency.

`sendWithMediaPromisedReply` could do that maybe

jernoble commented on Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
```diff
@@ -349,15 +349,17 @@ void SourceBufferPrivateRemote::setAppendWindowEnd(const MediaTime& appendWindow
+        return ComputeSeekPromise::createAndReject(-1);
 
-    gpuProcessConnection->connection().sendWithAsyncReply(Messages::RemoteSourceBufferProxy::ComputeSeekTime(target), WTFMove(completionHandler), m_remoteSourceBufferIdentifier);
+    return gpuProcessConnection->connection().sendWithPromisedReply(Messages::RemoteSourceBufferProxy::ComputeSeekTime(target), m_remoteSourceBufferIdentifier)->whenSettled(RunLoop::main(), [](auto&& result) {
```
Ditto here. For a result that's an `optional<>`, it could create or reject based on whether the optional has a value.

jyavenard replied:
the result is an `Expected<T, U>` The IPC Promise wraps the type set in the messages.in file, and an IPCError as exception.

It forces you to deal with the error if there's one.

jyavenard committed:
Use NativePromise chaining in place of operations queue in SourceBufferPrivate
https://bugs.webkit.org/show_bug.cgi?id=264854
rdar://118429088

Reviewed by Jer Noble.

By using promise chaining we can remove the need for an operations queue to serialise the tasks.
It also makes the code easier to read/follow as the entire appendBuffer operation is now available at a
glance.
This will simplify making SourceBufferPrivate::append/removeCodedFrames be made asynchronous and use
NativePromise.

* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::didUpdateFormatDescriptionForTrackId):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::processPendingMediaSamples):
(WebCore::SourceBufferPrivate::processMediaSample):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::~SourceBufferPrivate): Deleted.
(WebCore::SourceBufferPrivate::advanceOperationState): Deleted.
(WebCore::SourceBufferPrivate::rewindOperationState): Deleted.
(WebCore::SourceBufferPrivate::processAppendCompletedOperation): Deleted.
(WebCore::SourceBufferPrivate::queueOperation): Deleted.
(WebCore::SourceBufferPrivate::processPendingOperations): Deleted.
(WebCore::SourceBufferPrivate::abortPendingOperations): Deleted.
(WebCore::SourceBufferPrivate::processError): Deleted.
(WebCore::SourceBufferPrivate::processInitOperation): Deleted.
(WebCore::SourceBufferPrivate::processMediaSamplesOperation): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::didReceiveAllPendingSamples): Test that the promise wasn't previously resolved.
fallout from bug 264847.
(WebCore::SourceBufferPrivateGStreamer::appendParsingFailed):

Canonical link: https://commits.webkit.org/270808@main
